rust/src/graph.rs

```diff
@@ -0,0 +1,435 @@
+            build_info: self,
+            graph_handle,
+        })
+    }
+
+    #[inline(always)]
+    pub fn build_from_files<P>(self, files: impl AsRef<[P]>) -> Result<Graph, Error>
+    where
+        P: AsRef<std::path::Path>,
+    {
+        let mut graph_contents = Vec::with_capacity(files.as_ref().len());
+        for file in files.as_ref() {
+            graph_contents.push(std::fs::read(file).map_err(Into::<Error>::into)?);
+        }
+        let graph_builder_array: Vec<&[u8]> = graph_contents.iter().map(|s| s.as_ref()).collect();
+        let graph_handle =
+            syscall::load(graph_builder_array.as_slice(), self.encoding, self.target)?;
+        Ok(Graph {
+            build_info: self,
+            graph_handle,
+        })
+    }
+}
+
+/// An execution graph for performing inference (i.e., a model), which can create instances of [`GraphExecutionContext`].
+///
+/// ### Example
+/// ```rust
+/// use wasi_nn::{GraphBuilder, GraphEncoding, ExecutionTarget};
+///
+/// let model_file = "./test.tflite";
+/// // create a graph using `GraphBuilder`
+/// let graph = GraphBuilder::new(GraphEncoding::TensorflowLite, ExecutionTarget::CPU).build_from_files([model_file])?;
+/// // create an execution context using this graph
+/// let mut graph_exec_ctx = graph.init_execution_context()?;
+/// // set input tensors
+/// // ......
+///
+/// // compute the inference on the given inputs
+/// graph_exec_ctx.compute()?;
+///
+/// // get output tensors and do post-processing
+/// // ......
+/// ```
+pub struct Graph {
+    build_info: GraphBuilder,
+    graph_handle: syscall::GraphHandle,
+}
+
+impl Graph {
+    /// Get the graph encoding.
+    #[inline(always)]
+    pub fn encoding(&self) -> GraphEncoding {
+        self.build_info.encoding
+    }
+
+    /// Get the graph execution target.
+    #[inline(always)]
+    pub fn execution_target(&self) -> ExecutionTarget {
+        self.build_info.target
+    }
+
+    /// Get the graph id.
+    #[inline(always)]
+    pub fn graph_id(&self) -> syscall::GraphHandle {
+        self.graph_handle
+    }
+
+    /// Use this graph to create a new instances of [`GraphExecutionContext`].
+    #[inline(always)]
+    pub fn init_execution_context(&self) -> Result<GraphExecutionContext, Error> {
+        let ctx_handle = syscall::init_execution_context(self.graph_handle)?;
+        Ok(GraphExecutionContext {
+            graph: self,
+            ctx_handle,
+        })
+    }
+}
+
+/// Bind a [`Graph`] to the input and output for an inference.
+pub struct GraphExecutionContext<'a> {
+    graph: &'a Graph,
+    ctx_handle: syscall::GraphExecutionContextHandle,
+}
+
+impl<'a> GraphExecutionContext<'a> {
+    /// Get the [`Graph`] instance for this [`GraphExecutionContext`] instance.
+    #[inline(always)]
+    pub fn graph(&self) -> &Graph {
+        self.graph
+    }
+
+    /// Get the execution context id.
+    #[inline(always)]
+    pub fn context_id(&self) -> syscall::GraphExecutionContextHandle {
```

Comment:
Here and with `graph_id()`, I am debating whether we should even expose this to the user. If the user wants that kind of low-level detail, they can always use the raw, low-level API directly. What do you think about creating an `impl Debug` or `impl Display` that prints `GraphExecutionContext#...`? That way the information is there if someone wants to log it but we don't need to support these `*_id` functions forever if the underlying handle type changes.

Reply:
I saw it needs to print graph/context id in the example code, so I added the `*_id` functions.
Using the trait `Debug` and `Display` is a great idea, it avoids exposing low-level detail to the users.
What do you think we should print in debug/display, just a `*_id` or a formatted message such as "The Graph ID is `1`"?

Reply:
Let's do an `impl Display` for the context that prints `GraphExecutionContext#<id>` and another for graphs that prints `Graph#<id>` (are there any other types that contain IDs?). This still allows a user to log these things but it doesn't give the impression--like `impl Debug` would--that the `<id>` is the only thing contained in the struct.

Reply:
Ok. I have added them.